test_util/launch.py

```diff
@@ -1,20 +1,257 @@
+            config['parameteres'] = cf_params
+        return {
+            'delete_with_stack': generate_key,
+            'private_key': private_key,
+            'key_name': key_name,
+            'user': config.get('ssh_info', dict()).get('user', None)}
+
+    def test(self, info, test_cmd):
+        ssh_info = info['ssh']
+        for param in ['user', 'private_key']:
+            if param not in ssh_info:
+                raise LauncherError('MissingInfo', 'Cluster was created without providing ssh_info: {}'.format(param))
+        details = self.describe(info)
+        # FIX ME clean this up when done
+        test_host = details['masters'][0]['public_ip']
+        with ssh.tunnel.tunnel(ssh_info['user'], ssh_info['private_key'], test_host) as test_tunnel:
+            return test_util.runner.integration_test(
+                tunnel=test_tunnel,
+                dcos_dns=test_host,
+                master_list=[m['private_ip'] for m in details['masters']],
+                agent_list=[a['private_ip'] for a in details['private_agents']],
+                public_agent_list=[a['private_ip'] for a in details['public_agents']],
+                aws_access_key_id=info['provider']['access_key_id'],
+                aws_secret_access_key=info['provider']['secret_access_key'],
+                region=info['provider']['region'],
+                test_cmd=test_cmd)
+
+    def get_stack(self, info):
+        """Returns the correct class interface depending how the AWS CF is configured
+        NOTE: only supports Simple Cloudformation currently
+        """
+        try:
+            return DcosCfSimple(info['stack_name'], self.boto_wrapper)
+        except Exception as ex:
+            raise LauncherError('StackNotFound', '{} is not accessible'.format(info['stack_name'])) from ex
+
+
+def get_launcher(launcher_type, provider_info):
+    """Returns a launcher given a launcher type (string) and a dictionary of
+    appropriate provier_info
+    """
+    if launcher_type == 'cloudformation':
+        check_keys(provider_info, ['region', 'access_key_id', 'secret_access_key'])
+        boto_wrapper = BotoWrapper(
+            provider_info['region'], provider_info['access_key_id'], provider_info['secret_access_key'])
+        return AwsCloudformationLauncher(boto_wrapper)
+    else:
+        raise LauncherError('UnsupportedAction', 'Launcher type not supported: {}'.format(launcher_type))
+
+
+def convert_host_list(host_list):
+    """ Makes Host tuples more readable when using describe
+    """
+    return [{'private_ip': h.private_ip, 'public_ip': h.public_ip} for h in host_list]
+
+
+def check_keys(user_dict, key_list):
+    missing = [k for k in key_list if k not in user_dict]
+    if len(missing) > 0:
+        raise LauncherError('MissingInput', 'The following keys were required but'
+                            'not provided: {}'.format(repr(missing)))
+
+
+def main(argv):
+    args = docopt(__doc__, argv=argv, version='DC/OS Launch 1.0')
```

Comment:
Hm. I don't think its a bad idea to try to use one method of python arg handling, but I also feel its a arbitrary goal with little return-on-investment once we have a bunch of functional command line parsers already implemented. I.E. I think this belongs in the category of fixing something that is not broken.

As for the reason why I chose docopt: is that its documentation is perfectly in line with its function and boilerplate is absolutely minimal. The command line handling is not what I am concerned with in this PR; its the functionality that concerns me. If we decide we want this to have a more robust CLI, then we could look into that after we have created the launchers for all the providers

Furthermore, I don't really think there is such thing as a single python CLI that should be used. 

Docopt is the easiest to parse visually in the least amount of code. Ignore that example in the link you posted; its terrible. If you want to do fancy validation within the CLI, then just don't use docopt. Docopt is pointless if your final code looks that ugly IMO.

Argeparse is great if you just need to parse something extremely simple and want a script that can run on any machine with vanilla python and no modules.

Click is probably the 'best' option if you want fine-grained control. However, this will require a bunch of boiler plate and adding a new dependency to DC/OS (which is just one more unnecessary thing for now)